src/base/ClpMap.h

```diff
@@ -0,0 +1,248 @@
+    if (i == index.end()) {
+        return;
+    }
+
+    auto &e = (*i).second;
+    if (!expired(*e)) {
+        if (e != data.begin())
+            data.splice(data.begin(), data, e, std::next(e));
+        return;
+    }
+    // else fall through to cleanup
+
+    del(i);
+    i = index.end();
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+EntryValue *
+ClpMap<Key, EntryValue, MemoryUsedByEV>::get(const Key &key)
+{
+    KeyMapIterator i;
+    findEntry(key, i);
+    if (i != index.end()) {
+        const Entry &e = *(i->second);
+        return e.value;
+    }
+    return NULL;
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+size_t
+ClpMap<Key, EntryValue, MemoryUsedByEV>::memoryCountedFor(const Key &k, const EntryValue *v)
+{
+    // TODO: handle Entry which change size while stored
+    size_t entrySz = sizeof(Entry) + MemoryUsedByEV(v) + k.length();
+    return sizeof(MapItem) + k.length() + entrySz;
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+ClpMap<Key, EntryValue, MemoryUsedByEV>::add(const Key &key, EntryValue *t)
+{
+    return add(key, t, defaultTtl);
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+ClpMap<Key, EntryValue, MemoryUsedByEV>::add(const Key &key, EntryValue *t, int ttl)
+{
+    if (ttl == 0)
+        return false;
+
+    if (memLimit() == 0)
+        return false;
+
+    del(key);
+
+    const auto wantSz = memoryCountedFor(key, t);
+    if (wantSz >= memLimit())
+        return false;
+    trim(wantSz);
+
+    data.emplace_front(key, t, ttl);
+    index.emplace(key, data.begin());
+
+    ++entries_;
+    memUsed_ += wantSz;
+    return true;
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+ClpMap<Key, EntryValue, MemoryUsedByEV>::expired(const ClpMap::Entry &entry) const
+{
+    return entry.expires < squid_curtime;
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+ClpMap<Key, EntryValue, MemoryUsedByEV>::del(const KeyMapIterator &i)
+{
+    if (i != index.end()) {
+        auto &e = (*i).second;
+        const auto sz = memoryCountedFor(e->key, e->value);
+        data.erase(e);
+        index.erase(i);
+        --entries_;
+        memUsed_ -= sz;
+        return true;
+    }
+    return false;
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+bool
+ClpMap<Key, EntryValue, MemoryUsedByEV>::del(const Key &key)
+{
+    KeyMapIterator i;
+    findEntry(key, i);
+    return del(i);
+}
+
+template <class Key, class EntryValue, size_t MemoryUsedByEV(const EntryValue *)>
+void
+ClpMap<Key, EntryValue, MemoryUsedByEV>::trim(size_t wantSpace)
+{
+    while (memLimit() < (memoryUsed() + wantSpace)) {
+        auto i = data.end();
+        --i;
+        if (i != data.end()) {
+            del(i->key);
```

Comment:
Using rbegin() always produces an iterator. When the list is empty that iterator is rend().
The del() method uses find() and will always find the end iterator in the object (at the rend() position). Resulting in an infinite loop removing a non-existent entry.

We could use loop, but not avoid the complexity of checking the iterator value before deletion.
```suggestion
        auto i = data.rbegin();
        if (i != data.rend()) {
            del(i->key);
```

Reply:
> Using rbegin() always produces an iterator. When the list is empty that iterator is rend().

In the correct implementation, the list will never be empty in the middle of trim(). You can assert() that. There is no need to walk on tiptoes with iterators when the code is solid. It only hurts code understanding (and, in some cases, performance). Parts of the LruMap code were written in the "I am not sure what invariants I can rely on so I am going to check anything I can think of just in case something went wrong somewhere or I misunderstood something" style. We do not want to replicate that mistake in the new class.

BTW, _inlining_ trim()'s code, as I already suggested elsewhere, would make the "trim is guaranteed to succeed [because it must not be called otherwise]" principle more direct/obvious than using a method ever can, even if you describe the method with a beautiful 20-lline comment box documenting all the warnings and preconditions ;-).

Reply:
Okay. Done as of 5d500de6b now that the other code is actually maintaining those required memory invariants.